lib/llvm-backend/src/state.rs

```diff
@@ -231,33 +231,41 @@ impl<'ctx> State<'ctx> {
 
     pub fn outermost_frame(&self) -> Result<&ControlFrame<'ctx>, BinaryReaderError> {
         self.control_stack.get(0).ok_or(BinaryReaderError {
-            message: "invalid control stack depth",
+            message: "outermost_frame: invalid control stack depth",
             offset: -1isize as usize,
         })
     }
 
     pub fn frame_at_depth(&self, depth: u32) -> Result<&ControlFrame<'ctx>, BinaryReaderError> {
-        let index = self.control_stack.len() - 1 - (depth as usize);
-        self.control_stack.get(index).ok_or(BinaryReaderError {
-            message: "invalid control stack depth",
-            offset: -1isize as usize,
-        })
+        let index = self
+            .control_stack
+            .len()
+            .checked_sub(1 + (depth as usize))
```

Comment:
```
.checked_sub(1)
.checked_sub(depth as usize)
```
?

Reply:
Not possible because `checked_sub` return an `Option`.

Do you prefer this way?
``` rust
        let tmp_index = self
            .control_stack
            .len()
            .checked_sub(1).ok_or(BinaryReaderError {
                message: "frame_at_depth: invalid control stack depth",
                offset: -1isize as usize,
            })?;
        let index = tmp_index.checked_sub(depth as usize)
            .ok_or(BinaryReaderError {
                message: "frame_at_depth: invalid control stack depth",
                offset: -1isize as usize,
            })?;
        Ok(&self.control_stack[index])
```

Reply:
Doh! No, I prefer the code as it is, thanks.

Reply:
It should be possible to use `.and_then` here.  I noticed the potential for overflow there too, but I figured it's very unlikely.  I think it should be fine as-is!

For future readers who may end up here who didn't understand what I meant:

```Rust
self
            .control_stack
            .len()
            .checked_sub(1)
            .and_then(|n| n.checked_sub(depth as usize))
            .ok_or(BinaryReaderError {
                message: "frame_at_depth: invalid control stack depth",
                offset: -1isize as usize,
            })?;
```

`and_then` acts like Monadic bind (`>>=`) in Haskell.